### PR TITLE
Added support for hooks in the plan

### PR DESCRIPTION
**Pull request by ibragins**

### CODE REVIEW

File: ocp_resources/plan.py [hunk 1]
```diff
@@ -59,10 +63,39 @@ def __init__(
...
+        self.pre_hook_name = pre_hook_name
+        self.pre_hook_namespace = pre_hook_namespace
+        self.after_hook_name = after_hook_name
+        self.after_hook_namespace = after_hook_namespace
         self.target_namespace = target_namespace or self.namespace
         self.condition_message_ready = self.ConditionMessage.PLAN_READY
         self.condition_message_succeeded = self.ConditionMessage.PLAN_SUCCEEDED
 
+        hooks_array = []
+        if self.pre_hook_name and self.pre_hook_namespace:
```
Comment (myakove) [inline ocp_resources/plan.py:75]: can we have pre_hook without  after_hook ? is this a valid configuration?

**Comment by myakove** [comment]
Please fix CI failed tests (use pre-commit)

**Comment by myakove** [comment]
> /retest

You need to run `pre-commit install` (install it via pip if not installed already)
Then run `pre-commit run --all-files` (since you already pushed without the hook)
Next time you do `git commit` it will run automatically on the modified files.

**Comment by ibragins** [comment]
> > /retest
> 
> You need to run `pre-commit install` (install it via pip if not installed already) Then run `pre-commit run --all-files` (since you already pushed without the hook) Next time you do `git commit` it will run automatically on the modified files.

Thank you, did as you explained. Now we will see how it will work.

File: ocp_resources/hook.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
+from ocp_resources.mtv import MTV
+from ocp_resources.resource import NamespacedResource
+
+
+class Hook(NamespacedResource, MTV):
+    """
+    Migration Tool for Virtualization (MTV) Plan's Hook Resource.
+
+    Args:
```
Comment (rnetser) [inline ocp_resources/hook.py:9]: `Args` should be under `__init__`

File: ocp_resources/plan.py [hunk 3]
```diff
@@ -59,10 +63,39 @@ def __init__(
...
+            prehook = {
+                "hook": {
+                    "name": self.pre_hook_name,
+                    "namespace": self.pre_hook_namespace,
+                },
+                "step": "PreHook",
```
Comment (rnetser) [inline ocp_resources/plan.py:81]: Code is repeated for pre and post hook; please create a helper function which returns the required dict

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_resources/plan.py (hunk 3 above)
Comment (ibragins) [reply ocp_resources/plan.py]: Done, added function `get_hooks`

File: ocp_resources/hook.py (hunk 2 above)
Comment (ibragins) [reply ocp_resources/hook.py:9]: Please elaborate

Comment (ibragins) [reply ocp_resources/hook.py]: Done2

File: ocp_resources/plan.py [hunk 4]
```diff
@@ -59,10 +63,39 @@ def __init__(
...
         self.warm_migration = warm_migration
+        self.pre_hook_name = pre_hook_name
+        self.pre_hook_namespace = pre_hook_namespace
+        self.after_hook_name = after_hook_name
+        self.after_hook_namespace = after_hook_namespace
         self.target_namespace = target_namespace or self.namespace
         self.condition_message_ready = self.ConditionMessage.PLAN_READY
         self.condition_message_succeeded = self.ConditionMessage.PLAN_SUCCEEDED
 
+        def get_hook(hook_name, hook_namespace, hook_type):
```
Comment (rnetser) [inline ocp_resources/plan.py:74]: `get_hook` - please rename as this function does not get the hook but generates a dict. Something like generate_hook_spec

File: ocp_resources/plan.py [hunk 5]
```diff
@@ -59,10 +63,39 @@ def __init__(
...
+                    "namespace": hook_namespace,
+                },
+                "step": hook_type,
+            }
+            return hook
+
+        hooks_array = []
+        if self.pre_hook_name and self.pre_hook_namespace:
+            hooks_array.append(
+                get_hook(self.pre_hook_name, self.pre_hook_namespace, "PreHook")
```
Comment (rnetser) [inline ocp_resources/plan.py:87]: Please use are=val when calling a function:

```suggestion
                get_hook(hook_name=self.pre_hook_name, hook_namespace=self.pre_hook_namespace, hook_type="PreHook")
```

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_resources/plan.py (hunk 5 above)
Comment (ibragins) [reply ocp_resources/plan.py:87]: Done

File: ocp_resources/plan.py (hunk 4 above)
Comment (ibragins) [reply ocp_resources/plan.py:74]: Done

File: ocp_resources/hook.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        image="quay.io/konveyor/hook-runner:latest",
+        playbook=None,
+        client=None,
+        teardown=True,
+        yaml_file=None,
+    ):
+        """
+        Args:
+        image (str): Path to an ansible image
+        playbook (str): Ansible playbook to be performed
```
Comment (myakove) [inline ocp_resources/hook.py:25]: Missing indentation.

```
Args:
    image (str): Path to an ansible image
    playbook (str): Ansible playbook to be performed
```

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/hook.py (hunk 6 above)
Comment (ibragins) [reply ocp_resources/hook.py:25]: It was there, but pre-commit removed it

Comment (ibragins) [reply ocp_resources/hook.py]: fixed

**Pull request merged**